### ISSUE TITLE
[NEWS] GPSR - określ, czy produkt został wprowadzony na rynek UE przed wprowadzeniem rozporządzenia / GPSR - determine whether the product was placed on the EU market before the regulation was introduced

### ISSUE THREAD

**Issue by MaciejFrackowiak** [allegro/allegro-api#10548]
Od **13 grudnia 2024** obowiązuje nowe rozporządzenie Unii Europejskiej **[General Product Safety Regulation](https://commission.europa.eu/business-economy-euro/doing-business-eu/eu-product-safety-and-labelling/product-safety/general-product-safety-regulation_en)**. Celem GPSR jest pokazanie kupującemu, skąd pochodzi produkt, który może kupić na Allegro.

**Jak było dotychczas?**

Informację, czy produkt został wprowadzony do obrotu na terenie UE przed 13 grudnia 2024 można było przypisać do ofert, korzystając z [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers) podczas tworzenia oferty oraz [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers) przy jej edycji z wykorzystaniem parametru ofertowego o id=250553 “Produkt wprowadzony do obrotu na terenie UE przed 13.12.2024”.

**Co zmieniliśmy?**

Dzisiaj, dla produktów wprowadzonych na rynek UE przed wdrożeniem GPSR, udostępniliśmy nową flagę “**marketedBeforeGPSRObligation**” dla każdego produktu (**productSet.[].marketedBeforeGPSRObligation**) w zestawie (productSet).

**Wartość:**

* **true** - oznacza, że produkt został wprowadzony na rynek UE przed obowiązywaniem rozporządzenia GPSR;
* **false** - oznacza, że produkt został wprowadzony na rynek UE już po wdrożeniu rozporządzenia GPSR.

**Przykładowy request:**

```json5
curl -X PATCH \
'https://api.allegro.pl/sale/product-offers/7776506218' \
-H 'Authorization: Bearer {token}' \
-H 'Accept: application/vnd.allegro.public.v1+json' \
-H 'Content-Type: application/vnd.allegro.public.v1+json' \
-d '{
	"productSet": [
    	   {
        	"marketedBeforeGPSRObligation": true,
		...
    	   },
    	   {
        	"marketedBeforeGPSRObligation": false,     			
	        ...
    	   }
	]
	...
}'
```
> 

**Uwaga!** Edytując ofertę ([PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers)) w przypadku tablic (np. productSet) przyjmujemy zasadę **wszystko albo nic** (zgodnie z [RFC7396](https://tools.ietf.org/html/rfc7396)). Oznacza to, że jeśli chcesz np. ustawić flagę “**marketedBeforeGPSRObligation**” musisz przekazać jej wartość, jak i dotychczasowe dane opisujące zestaw produktowy (które zwracamy w tablicy productSet dla oferty).

Jednocześnie dzisiaj udostępniliśmy możliwość grupowej zmiany flagi “**marketedBeforeGPSRObligation**” z wykorzystaniem [PUT /sale/offer-modification-commands/{commandId}](https://developer.allegro.pl/documentation/#operation/modificationCommandUsingPUT).

**Przykładowy request:**

```json5
curl -X PUT \
'https://api.allegro.pl/sale/offer-modification-commands/{commandId}' \
-H 'Authorization: Bearer {token}' \
-H 'Accept: application/vnd.allegro.public.v1+json' \
-H 'Content-Type: application/vnd.allegro.public.v1+json' \
-d '{
	"modification": {
    	    "marketedBeforeGPSRObligation": true  // informacja, czy produkt został wprowadzony na rynek UE przed wdrożeniem GPSR
	},
	"offerCriteria": [		
    	    {
        	   "offers": [ 		               // lista ofert, w których chcesz dokonać zmian
            	  {
                	    "id": "11223344556"
            	  },
            	  {
                	    "id": "11335577991"
            	  }
        	   ],
        	   "type": "CONTAINS_OFFERS"
    	    }
	]
}'
```
> 

**Uwaga!** Grupowo flagę “**marketedBeforeGPSRObligation**“ zmienisz **tylko** dla **pierwszego** produktu w zestawie (productSet).

**Dlaczego wprowadziliśmy zmianę?**

Zależy nam, aby sprzedający mogli w możliwie najwygodniejszy dla siebie sposób określić, czy sprzedawane przez nich produkty zostały wprowadzone na rynek UE przed dniem wdrożenia GPSR.

-----
As of **December 13, 2024**, the new European Union Regulation **[General Product Safety Regulation](https://commission.europa.eu/business-economy-euro/doing-business-eu/eu-product-safety-and-labelling/product-safety/general-product-safety-regulation_en)** is in effect. The goal of the GPSR is to show the buyer where the product available on Allegro originates.

**How was it so far?**

Information on whether a product was marketed in the EU before December 13, 2024, could be assigned to offers using [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers) when creating an offer and [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers) when editing it using the offer parameter with id=250553 “Product introduced to the EU market before 13.12.2024”.

**What have we changed?**

Today, for products launched in the EU before the implementation of GPSR, we have provided a new “**marketedBeforeGPSRObligation**” flag for each product (**productSet.[].marketedBeforeGPSRObligation**) in the set (productSet).

The value:

* **true** - means that the product was placed on the EU market before the GPSR regulation was in force;
* **false** - means that the product has already been placed on the EU market after the implementation of the GPSR regulation.

**Sample request:**

```json5
curl -X PATCH \
'https://api.allegro.pl/sale/product-offers/7776506218' \
-H 'Authorization: Bearer {token}' \
-H 'Accept: application/vnd.allegro.public.v1+json' \
-H 'Content-Type: application/vnd.allegro.public.v1+json' \
-d '{
	"productSet": [
    	   {
        	"marketedBeforeGPSRObligation": true,
                ...
    	   },
    	   {
        	"marketedBeforeGPSRObligation": false,
                ...
    	   }
	]
	...
}'
```
> 

**Note!** When editing an offer ([PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers)) for arrays (e.g. productSet), we adopt the **all-or-nothing** rule (according to [RFC7396](https://tools.ietf.org/html/rfc7396)). This means that if you want, for example, to set the flag (“**marketedBeforeGPSRObligation**”) you must pass their value, as well as the existing data describing the product set (which we return in the productSet array for the offer).

At the same time, today we have provided the ability to batch change the “**marketedBeforeGPSRObligation**” flag using [PUT /sale/offer-modification-commands/{commandId}](https://developer.allegro.pl/documentation/#operation/modificationCommandUsingPUT).

**Sample request:**

```json5
curl -X PUT \
'https://api.allegro.pl/sale/offer-modification-commands/{commandId}' \
-H 'Authorization: Bearer {token}' \
-H 'Accept: application/vnd.allegro.public.v1+json' \
-H 'Content-Type: application/vnd.allegro.public.v1+json' \
-d '{
	"modification": {
    	    "marketedBeforeGPSRObligation": true  // information on whether the product was launched in the EU market before the implementation of GPSR
	},
	"offerCriteria": [		
    	    {
        	   "offers": [ 		               // list of offers you want to edit
            	  {
                	    "id": "11223344556"
            	  },
            	  {
                	    "id": "11335577991"
            	  }
        	   ],
        	   "type": "CONTAINS_OFFERS"
    	    }
	]
}'
```
> 

**Note!** You will batch change the flag “**marketedBeforeGPSRObligation**“ **only** for the **first** product in the set (productSet).

**Why have we made the change?**

We want sellers to be able to determine as conveniently as possible whether the products they sell were placed on the EU market before the date of GPSR implementation.

**Comment by dominikahr** [comment]
Just to be sure, we understand this correctly: safety information and responsible producer are mandatory parameters in all categories, that fall under the GPSR regulation. The merchant can also select the option "marketedBeforeGPSRObligation": true and then the merchant will NOT have to fill out the safety information and responsible producer?

**Comment by jezewskiBNetrox** [comment]
@dominikahr Yes. This works as product-wise version of offer parameter they added earlier. Parameter however will be removed as it is temporary solution that will cease to be.

**Comment by MaciejFrackowiak** [comment]
@dominikahr I can confirm that it works exactly as @jezewskiBNetrox described.
